staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width_test.go

```diff
@@ -0,0 +1,180 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"net/http"
+	"testing"
+
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+func TestWidthEstimator(t *testing.T) {
+	tests := []struct {
+		name          string
+		requestURI    string
+		requestInfo   *apirequest.RequestInfo
+		counts        map[string]int64
+		seatsExpected uint
+	}{
+		{
+			name:          "request has no RequestInfo",
+			requestURI:    "http://server/apis/v1/foos/",
+			requestInfo:   nil,
+			seatsExpected: 10,
+		},
+		{
+			name:       "request verb is not list",
+			requestURI: "http://server/apis/v1/foos/",
+			requestInfo: &apirequest.RequestInfo{
+				Verb: "get",
+			},
+			seatsExpected: 1,
+		},
+		{
+			name:       "request verb is list, resource version not set",
+			requestURI: "http://server/apis/v1/foos/1?limit=499",
+			requestInfo: &apirequest.RequestInfo{
+				Verb:     "list",
+				APIGroup: "foo.bar",
+				Resource: "resource",
+			},
+			counts: map[string]int64{
+				"resource.foo.bar": 1999,
+			},
+			seatsExpected: 5,
+		},
+		{
+			name:       "request verb is list, continuation is set",
+			requestURI: "http://server/apis/v1/foos/1?continue=token&limit=499&resourceVersion=1",
+			requestInfo: &apirequest.RequestInfo{
+				Verb:     "list",
+				APIGroup: "foo.bar",
+				Resource: "resource",
+			},
+			counts: map[string]int64{
+				"resource.foo.bar": 1999,
+			},
+			seatsExpected: 5,
+		},
+		{
+			name:       "request verb is list, has limit",
+			requestURI: "http://server/apis/v1/foos/1?limit=499&resourceVersion=1",
+			requestInfo: &apirequest.RequestInfo{
+				Verb:     "list",
+				APIGroup: "foo.bar",
+				Resource: "resource",
+			},
+			counts: map[string]int64{
+				"resource.foo.bar": 1999,
+			},
+			seatsExpected: 5,
+		},
+		{
+			name:       "request verb is list, resource version is zero",
+			requestURI: "http://server/apis/v1/foos/1?resourceVersion=0",
+			requestInfo: &apirequest.RequestInfo{
+				Verb:     "list",
+				APIGroup: "foo.bar",
+				Resource: "resource",
+			},
+			counts: map[string]int64{
+				"resource.foo.bar": 799,
+			},
+			seatsExpected: 8,
+		},
+		{
+			name:       "request verb is list, no query parameters",
+			requestURI: "http://server/apis/v1/foos/1",
+			requestInfo: &apirequest.RequestInfo{
+				Verb:     "list",
+				APIGroup: "foo.bar",
+				Resource: "resource",
+			},
+			counts: map[string]int64{
+				"resource.foo.bar": 799,
+			},
+			seatsExpected: 1,
```

Comment:
Hmm - with no query params, we should be delegating to storage. Which means we will be using limit.

But with limit unsed, we're using limit=0, which results in 1 here.
That sounds a bit wrong.

We should update the logic to:
- if the limit is 0, fallback to <count> (and if that is also 0, then use max)

Reply:
sounds good, i have added fallback logic, please take a look.